src/Runner/Runner.php

```diff
@@ -89,26 +119,245 @@ public function __construct(
+                $process->bindConnection($decoder, $encoder);
+                $job = $fileChunk();
+
+                if (0 === \count($job)) {
+                    $process->request(['action' => ParallelAction::WORKER_THANK_YOU]);
+                    $processPool->endProcessIfKnown($identifier);
+
+                    return;
+                }
+
+                $process->request(['action' => ParallelAction::WORKER_RUN, 'files' => $job]);
+            });
+        });
+
+        $processesToSpawn = min(
+            $this->parallelConfig->getMaxProcesses(),
+            (int) ceil($this->fileCount / $this->parallelConfig->getFilesPerProcess())
         );
+        $processFactory = new ProcessFactory($this->input);
+
+        for ($i = 0; $i < $processesToSpawn; ++$i) {
+            $identifier = ProcessIdentifier::create();
+            $process = $processFactory->create(
+                $streamSelectLoop,
+                new RunnerConfig(
+                    $this->isDryRun,
+                    $this->stopOnViolation,
+                    $this->parallelConfig,
+                    $this->configFile
+                ),
+                $identifier,
+                $serverPort,
+            );
+            $processPool->addProcess($identifier, $process);
+            $process->start(
+                // [REACT] Handle workers' responses (multiple actions possible)
+                function (array $workerResponse) use ($processPool, $process, $identifier, $fileChunk, &$changed): void {
+                    // File analysis result (we want close-to-realtime progress with frequent cache savings)
+                    if (ParallelAction::RUNNER_RESULT === $workerResponse['action']) {
+                        $fileAbsolutePath = $workerResponse['file'];
+                        $fileRelativePath = $this->directory->getRelativePathTo($fileAbsolutePath);
+
+                        // Pass-back information about applied changes (only if there are any)
+                        if (isset($workerResponse['fixInfo'])) {
+                            $changed[$fileRelativePath] = $workerResponse['fixInfo'];
+                        }
+                        // Dispatch an event for each file processed and dispatch its status (required for progress output)
+                        $this->dispatchEvent(FixerFileProcessedEvent::NAME, new FixerFileProcessedEvent($workerResponse['status']));
+
+                        if (
+                            FixerFileProcessedEvent::STATUS_NO_CHANGES === (int) $workerResponse['status']
+                            || (
+                                FixerFileProcessedEvent::STATUS_FIXED === (int) $workerResponse['status']
+                                && !$this->isDryRun
+                            )
+                        ) {
+                            $this->cacheManager->setFile($fileRelativePath, file_get_contents($fileAbsolutePath));
```

Comment:
in `sequential` flow, we read file once and in setting the cache we put new code base on code from tokens (no extra read operation)

what I see here is that we read same file in Worker command and in Runner. I would love to avoid that [eg by setting file by Worker command]

Reply:
I think I don't understand this concern. In parallel run cache file is _read_ once per process, because every worker needs that info. Cache _write_ is done only in the main process, after worker reports the result.

Reply:
how I understand the code


sequential flow:
- `file_get_contents` is called
- file is fixed
- `file_put_contents` is called with result of `$new = $tokens->generateCode()`
- `$cacheManager->setFile($file, $new)` is called

parallel flow:
- `file_get_contents` is called
- file is fixed
- `file_put_contents` is called with result of `$tokens->generateCode()` ## worker
- `$cacheManager->setFile($file, file_get_contents($file))` is called ## supervisor
  - this is place of extra call to `file_get_contents`

Reply:
Ah, I get it now. Yeah, since the cache is written in the main process, we would need to pass whole new content via socket, which can affect the performance and has a risk of reaching buffer's size. We can't write the cache in worker because of potential race conditions. That's why we need an additional file read.

Reply:
I do not believe we should ignore this concern. You are describing why we do what we do, not how to avoid that.

while I do not have the solution for this concern as only trying to understand this PR for now, perhaps this can be option to go?
- eg by setting file by Worker command (suggestion from first msg in this thread)
  - I hear you about race condition. there is a cache-file locking in place, thus cache shouldn't be corrupted (but it can introduce wait time, agree)
- `workerResponse` could include not only `fixInfo`, but also checksum ready-to-use for cache

Reply:
> by setting file by Worker command

It was like that before and it was causing cache issues - multiple runs were required to fully fill the cache. That's why it was moved to main process, which handles cache saving one-by-one. More info was provided [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7777#discussion_r1482820415).

> workerResponse could include not only fixInfo, but also checksum ready-to-use for cache

It does not solve the issue because checksum is an internal implementation detail of `FileCacheManager`. The contract is based on file contents (`PhpCsFixer\Cache\CacheManagerInterface::setFile(string $file, string $fileContent): void`) which requires either sending fixed file's content via socket (I prefer to avoid it, as pointed before), or reading it again in the main process. I don't plan changing `CacheManagerInterface` under this PR, I believe it is good as-is and we can improve it later in separate PR, which will be easier to review.